### PR TITLE
Finished with oop exercise

### PR DESCRIPTION
**Pull request by kalash14**
_Complete the form below and remove this line._

### OOP exercise

[Demo](https://kalash14.github.io/oop-exercise/) |
[Code base](https://github.com/kalash14/oop-exercise/blob/master/index.js)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

**Comment by OleksiyRudenko** [comment]
@kalash14 Looks good at a glance. Please, meanwhile check yourself against the following list:

 * [ ] Implement a base class to inherit from
 * [ ] Employ default parameters
 * [ ] Each species is represented with its own class
 * [ ] No need to specify species at instantiation
 * [ ] Classes for species that do not have hands by natural design
       do not consequently have `hands` or any equivalent property
 * [ ] All inhabitants are stored in a container (array or object)
 * [ ] JS native features are intensively employed (`const`, `let`, `Array.map|join|forEach|...`, etc)
 * [ ] [OOP, SOLID and DRY](https://github.com/OleksiyRudenko/a-tiny-JS-world/blob/master/README.md#learn-on-your-own)
       principles are intensively employed
 
Optional level up (not required to implement):
 * [ ] Friends list is a list of objects refs rather than names (strings)
 * [ ] Cat-woman class is built employing composition rather than inheritance only
 
[ES6 classes cheat-sheet](https://gist.github.com/OleksiyRudenko/672d39b08d9d0da4e179aca49876c58b)

Please, confirm all items above are positive. Please, ask a question if anything above requires clarification.

**Comment by OleksiyRudenko** [comment]
@kalash14 Please, let me know somehow that you have implemented everything you wanted to and also 
> Please, confirm all items above are positive. Please, ask a question if anything above requires clarification.

**Comment by kalash14** [comment]
In my previous commit, I have reworked my code according to your recommendations and I have implemented the next items:

- [x] Implement a base class to inherit from
- [x] Employ default parameters
- [x] Each species is represented with its own class
- [x] No need to specify species at instantiation
- [x] Classes for species that do not have hands by natural design
do not consequently have hands or any equivalent property
- [x] All inhabitants are stored in a container (array or object)
- [x] JS native features are intensively employed (const, let, Array.map|join|forEach|..., etc)
- [x] OOP, SOLID and DRY principles are intensively employed 

AFAIU, I should make a commit after every implemented item from the list above. Right? Thanks for your answer.

**Comment by kalash14** [comment]
Also, to my mind, in order to follow DRY principles, I would like to create an instance 'parameters' and put there all parameters of the base class, because I repeated them in my code and I think it's not OK

**Comment by OleksiyRudenko** [comment]
> AFAIU, I should make a commit after every implemented item from the list above. Right? Thanks for your answer.

It is up to you. Just let me know when you finish with code improvements or have a question.

**Comment by OleksiyRudenko** [comment]
> Also, to my mind, in order to follow DRY principles, I would like to create an instance 'parameters' and put there all parameters of the base class, because I repeated them in my code and I think it's not OK

Good catch ;)

Whatever makes code DRY is good.

**Comment by kalash14** [comment]
I have finished with code improvements.

File: submissions/kalash14/oop-exercise/index.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    introduceCreature () {
+        return `<p>${this.greetingsWord}!!! Nice to meet you. My name is ${this.name} and I'm ${this.kind}. Also, I'm a ${this.gender}. ${(this.legs ? `I have ${this.legs} legs.` : '')} My favourite dish is ${this.favouriteDish}. </p>`;
+    }
+
+}
+
+class Human extends Inhabitant {
+
+    constructor(name, gender, greetingsWord, kind, favouriteDish, legs = 2) {
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: Do you really want to let the caller to provide `Human` with any kind at caller's discretion?

File: submissions/kalash14/oop-exercise/index.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+	constructor(name, gender, greetingsWord, kind, favouriteDish, legs = 4) {
+        super(name, gender, greetingsWord, kind, favouriteDish);
+        this.legs = legs;
+	}
+
+}
+
+class Dog extends Animal {
+    constructor(name, gender, greetingsWord = 'Bow-wow', kind = 'dog', favouriteDish, legs) {
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: Do you really want to let the caller to provide `Dog` with any kind at caller's discretion?
Do you really want to let the caller to provide `Dog` with any number of legs at caller's discretion?

Ask same question on every parameter of every class constructors. Respond wisely and fix correspondingly.

File: submissions/kalash14/oop-exercise/index.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    constructor(name, gender, greetingsWord, kind, favouriteDish, legs) {
+        super(name, gender, greetingsWord, kind, favouriteDish, legs);
+    }
+}
+
+const inhabitantsArray = [
+
+    new Dog('Sharik', 'male', 'Bow-wow', 'dog', 'meat'),
+    new Cat('Kuzia', 'male', 'Meow-meow', 'cat', 'boiled fish'),
+    new Woman('Leila', 'female', 'Hi', 'woman', 'Cesar salad'),
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: Please, note that `woman` and `man` are both of human species.

File: submissions/kalash14/oop-exercise/index.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    new Man('Pedro Rodriguez', 'male', 'Hey, dude', 'man', 'Avocado sandwich')
+
+];
+
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: Please, read carefully docs with regards to output and fix the code below correspondingly.

File: submissions/kalash14/oop-exercise/index.js (hunk 1 above)
Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: Oops...

File: submissions/kalash14/oop-exercise/index.js (hunk 2 above)
Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: Ok

File: submissions/kalash14/oop-exercise/index.js (hunk 3 above)
Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: Ok, thanks

File: submissions/kalash14/oop-exercise/index.js (hunk 4 above)
Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: Ok

File: submissions/kalash14/oop-exercise/index.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+   Code repository: https://github.com/kalash14/oop-exercise/blob/populate-world/index.js
+   Web app: https://kalash14.github.io/oop-exercise/
+   */
+
+// ======== OBJECTS DEFINITIONS ========
+
+class Inhabitant {
+
+    constructor(name, gender) {
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: Please, initialize in `Inhabitant` all the properties that are common for all inhabitants.

Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: Common properties for all inhabitants: name, gender, kind and greetingsWords

Comment (OleksiyRudenko) [reply submissions/kalash14/oop-exercise/index.js]: @kalash14 That's correct. Please, fix `Inhabitant`.

Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: Ok

**Comment by OleksiyRudenko** [comment]
@kalash14 Well, I hoped we can complete the task in 3 steps.
Please, compare your own suggestion 
> Common properties for all inhabitants are name, gender, kind, greetingsWords and legs
with your current code.

Please, take comments as verbose technical requirements documentation.

**Comment by kalash14** [comment]
I noticed that I didn't inherit these properties in each class

**Comment by kalash14** [comment]
Thank you for your patience and sorry for my inattention. I have fixed these parameters.

File: submissions/kalash14/oop-exercise/index.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        this.gender = gender;
+        this.kind = kind;
+        this.greetingsWords = greetingsWords;
+        this.legs = legs;
+    }
+}
+
+class Human extends Inhabitant{
+
+    constructor(name, gender, greetingsWords, legs = 2) {
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: You can ask user for hands as an argument to `new Human` with a default value like `legs`.

File: submissions/kalash14/oop-exercise/index.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        super(name, gender, 'Human', greetingsWords, legs);
+        this.hands = 2;
+        this.greetingsWords = `Hi!`
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: Why do you override `greetingsWords`?
You receive something from a user and pass to the base class via `super`, so the property gets initialized properly.

File: submissions/kalash14/oop-exercise/index.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+    constructor(name, gender, kind, greetingsWords, legs = 4) {
+        super(name, gender, kind, greetingsWords, legs);
+    }
+
+}
+
+class Dog extends Animal {
+
+    constructor(name, gender, greetingsWords, legs = 4) {
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: If you want a dog to have a standard greeting then you do need to ask a user for one.

File: submissions/kalash14/oop-exercise/index.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        super(name, gender, 'Dog', greetingsWords, legs);
+        this.greetingsWords = 'Bow-wow!';
+    }
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: And when calling `super` you can specify the default greeting explicitly.
Super class will set the property for you.

File: submissions/kalash14/oop-exercise/index.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+}
+
+class Cat extends Animal {
+
+    constructor(name, gender, greetingsWords, legs = 4) {
+        super(name, gender, 'Cat', greetingsWords, legs);
+        this.greetingsWords = 'Meow!';
+    }
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: Same as for `Dog`.

File: submissions/kalash14/oop-exercise/index.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+}
+
+const creatures = [
+
+    new Human('Victor', 'male'),
+    new Human('Helen', 'female'),
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: Let humans have their individual greetings.

File: submissions/kalash14/oop-exercise/index.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    new Cat('Pushok', 'male'),
+    new Dog('Mukhtar', 'male')
+
+];
+
+creatures.forEach(creature => print(creature.kind + "; " + creature.gender + "; " + creature.name + "; " + creature.legs + "; " + (creature.hands ? creature.hands + "; " : "") + creature.greetingsWords));
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: Please, add to the basemost class a method that would return a nicely formatted string. That's how separation of concerns principle gets employed.

Then here run through array and call that method for every creature within `print()`.

File: submissions/kalash14/oop-exercise/index.js (hunk 6 above)
Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: ok!

File: submissions/kalash14/oop-exercise/index.js (hunk 7 above)
Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: ok!

File: submissions/kalash14/oop-exercise/index.js (hunk 8 above)
Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: ok!

File: submissions/kalash14/oop-exercise/index.js (hunk 10 above)
Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: ok!

File: submissions/kalash14/oop-exercise/index.js (hunk 9 above)
Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: ok!

File: submissions/kalash14/oop-exercise/index.js (hunk 11 above)
Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: ok!

File: submissions/kalash14/oop-exercise/index.js (hunk 12 above)
Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: ok!

File: submissions/kalash14/oop-exercise/index.js [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+class Human extends Inhabitant{
+
+    constructor(name, gender, greetingsWords, legs = 2, hands = 2) {
+        super(name, gender, 'Human', greetingsWords, legs);
+        this.hands = hands;
+    }
+
+}
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: In method overloading we trust.
Define `greetingMessage` for `Human`. Use `greetingMessage` from super class (`super.greetingMessage()`) and add something about hands. No problem that hands come after everything else.

BTW, as you might remember function names should start with a verb.

As you have to rename it anyway rename `greetingMessage` to `toString` everywhere. Check for the bonus learning below.

File: submissions/kalash14/oop-exercise/index.js [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+const creatures = [
+
+    new Human('Victor', 'male' , 'Hello!'),
+    new Human('Helen', 'female', 'Hi!'),
+    new Cat('Pushok', 'male'),
+    new Dog('Mukhtar', 'male')
+
+];
+
+creatures.forEach(creature => print(creature.greetingMessage()));
```
Comment (OleksiyRudenko) [inline submissions/kalash14/oop-exercise/index.js]: If your class (object) has magic method `toString` you can benefit from it in string contexts.
`print(creature)` will call `.toString()` implicitly for you.

Comment (OleksiyRudenko) [reply submissions/kalash14/oop-exercise/index.js]: Please, ensure your files have an empty line at the end. It's important. You can tune your IDE so it ensures empty lines for you.

Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: Wow! Cool! 
Ok, I have already configured my IDE.

File: submissions/kalash14/oop-exercise/index.js (hunk 13 above)
Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: Ok!

File: submissions/kalash14/oop-exercise/index.js [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+        this.name = name;
+        this.gender = gender;
+        this.kind = kind;
+        this.greetingsWords = greetingsWords;
+        this.legs = legs;
+    }
+
+    greetingMessage() {
+        return `${this.kind}; ${this.gender}; ${this.name}; ${this.legs}; ${(this.hands ? this.hands + "; " : "")}${this.greetingsWords}`;
+    }
```
[the remark this answers is not in the document]
Comment (kalash14) [reply submissions/kalash14/oop-exercise/index.js]: Ok!

**Comment by kalash14** [comment]
Thanks for all recommendations! Happy New year and Merry Christmas!!!

**Pull request merged**